cpp/crawl_v1_executor.cpp

```diff
@@ -0,0 +1,163 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <iostream>
+#include <vector>
+
+#include <boost/regex.hpp>
+
+#include <curl/curl.h>
+
+#include <stout/os.hpp>
+
+#include "rendler_helper.hpp"
+#include "rendler_v1_executor.hpp"
+
+using std::cout;
+using std::endl;
+using std::vector;
+
+using mesos::vectorToString;
+
+
+static int writer(char *data, size_t size, size_t nmemb, string *writerData)
+{
+  assert(writerData != NULL);
+  writerData->append(data, size*nmemb);
+  return size * nmemb;
+}
+
+
+class CrawlV1Executor : public RendlerV1Executor
+{
+public:
+  CrawlV1Executor(const FrameworkID& _frameworkId,
+                  const ExecutorID& _executorId)
+    : RendlerV1Executor("CrawlV1Executor", _frameworkId, _executorId) {}
+
+  virtual ~CrawlV1Executor() {}
+
+protected:
+  void runTask(const TaskInfo& task) override
+  {
+    string url = task.data();
+    cout << "Running crawl task " << task.task_id().value()
+         << " Fetch: " << url;
+
+    string buffer;
+    vector<string> result;
+    result.push_back(task.task_id().value());
+    result.push_back(url);
+
+    CURL *conn;
+    conn = curl_easy_init();
+    assert(conn != NULL);
+    assert(curl_easy_setopt(conn, CURLOPT_URL, url.c_str()) == CURLE_OK);
+    assert(curl_easy_setopt(conn, CURLOPT_FOLLOWLOCATION, 1L) == CURLE_OK);
+    assert(curl_easy_setopt(conn, CURLOPT_WRITEFUNCTION, writer) == CURLE_OK);
+    assert(curl_easy_setopt(conn, CURLOPT_WRITEDATA, &buffer) == CURLE_OK);
+
+    if (curl_easy_perform(conn) != CURLE_OK) {
+      return;
+    }
+
+    char *tmp;
+    assert(curl_easy_getinfo(conn, CURLINFO_EFFECTIVE_URL, &tmp) == CURLE_OK);
+    string redirectUrl = url;
+    if (tmp != NULL) {
+      redirectUrl = tmp;
+    }
+    curl_easy_cleanup(conn);
+
+    size_t scheme = redirectUrl.find_first_of("://");
+    size_t sp = redirectUrl.find_first_of('/',
+        scheme == string::npos ? 0 : scheme + 3); // skip the http:// part.
+    size_t lsp = redirectUrl.find_last_of('/'); // skip the http:// part.
+    string baseUrl = redirectUrl.substr(0, sp); // No trailing slash.
+    string dirUrl = redirectUrl.substr(0, lsp); // No trailing slash.
+
+    cout << "redirectUrl " << redirectUrl << " baseURL: " << baseUrl << endl;
+    cout << "dirUrl " << dirUrl  << endl;
+
+    const boost::regex hrefRE("<a\\s+[^\\>]*?href\\s*=\\s*([\"'])(.*?)\\1");
+    const boost::regex urlRE("^([a-zA-Z]+://).*");
+
+    boost::smatch matchHref;
+    string::const_iterator f = buffer.begin();
+    string::const_iterator l = buffer.end();
+
+    while (f != buffer.end() &&
```

Comment:
I think it has an overload for a string as well?
http://www.boost.org/doc/libs/1_61_0/libs/regex/doc/html/boost_regex/ref/regex_search.html